kubernetes/config/kube_config.c

```diff
@@ -0,0 +1,309 @@
+#include "kube_config.h"
+#include "kube_config_yaml.h"
+#include <errno.h>
+#include <unistd.h>
+
+#define ENV_KUBECONFIGE "KUBECONFIG"
+#define ENV_HOME "HOME"
+#define KUBE_CONFIG_DEFAULT_LOCATION "%s/.kube/config"
+#define KUBE_CONFIG_TEMPFILE_NAME_TEMPLATE "/tmp/kubeconfig-XXXXXX"
+
+static int setBasePath(char **pBasePath, char *basePath)
+{
+    int rc = 0;
+    char *_basePath = strdup(basePath);
+    if (_basePath) {
+        *pBasePath = _basePath;
+    } else {
+        rc = -1;
+    }
+    return rc;
+
+}
+
+static char *kubeconfig_mk_cert_key_tempfile(const char *b64data)
+{
+    static char fname[] = "kubeconfig_mk_tempfile()";
+
+    int decoded_bytes = 0;
+    char *b64decode = base64decode(b64data, strlen(b64data), &decoded_bytes);
+    if (!b64decode || 0 == decoded_bytes) {
+        fprintf(stderr, "%s: Base64 decodes failed.\n", fname);
+        return NULL;
+    }
+
+    char tempfile_name_template[] = KUBE_CONFIG_TEMPFILE_NAME_TEMPLATE;
+    int fd = mkstemp(tempfile_name_template);
+    if (-1 == fd) {
+        fprintf(stderr, "%s: Creating temp file for kubeconfig failed with error [%s]\n", fname, strerror(errno));
+        return NULL;
+    }
+
+    if (-1 == write(fd, b64decode, decoded_bytes)) {
+        fprintf(stderr, "%s: Writing temp file failed with error [%s]\n", fname, strerror(errno));
+        close(fd);
```

Comment:
hrm, I kind of don't love the `close(fd)` in multiple places, it's vulnerable to later cut/paste bugs.

How about:

```c
int rc = write(...);
close(fd);
return (rc == -1 ? NULL : strdup(...));
```

Reply:
OK.